homeassistant/components/ssdp/__init__.py

```diff
@@ -43,134 +48,203 @@
+    return scanner.async_register_callback(callback, match_dict)
+
+
+@bind_hass
+def async_get_location_by_udn_st(  # pylint: disable=invalid-name
+    hass: HomeAssistant, udn: str, st: str
+) -> dict[str, str] | None:
+    """Lookup a location by udn and st from previous scans."""
+    scanner: Scanner = hass.data[DOMAIN]
+    return scanner.cache.get((udn, st))
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the SSDP integration."""
 
-    async def _async_initialize(_):
-        scanner = Scanner(hass, await async_get_ssdp(hass))
-        await scanner.async_scan(None)
-        cancel_scan = async_track_time_interval(hass, scanner.async_scan, SCAN_INTERVAL)
+    scanner = hass.data[DOMAIN] = Scanner(hass, await async_get_ssdp(hass))
 
-        @callback
-        def _async_stop_scans(event):
-            cancel_scan()
+    asyncio.create_task(scanner.async_start())
 
-        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_stop_scans)
+    return True
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, _async_initialize)
 
+def _async_use_default_interface(adapters: list[network.Adapter]) -> bool:
+    for adapter in adapters:
+        if adapter["enabled"] and not adapter["default"]:
+            return False
     return True
 
 
+def _callback_if_match(
+    callback: Callable[[dict], None], info: dict[str, str], match_dict: dict[str, str]
+) -> None:
+    """Fire a callback if info matches the match dict."""
+    if not all(item in info.items() for item in match_dict.items()):
+        return
+    try:
+        callback(info)
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("Failed to callback info: %s", info)
+
+
 class Scanner:
     """Class to manage SSDP scanning."""
 
-    def __init__(self, hass, integration_matchers):
+    def __init__(
+        self, hass: HomeAssistant, integration_matchers: dict[str, list[dict[str, str]]]
+    ) -> None:
         """Initialize class."""
         self.hass = hass
-        self.seen = set()
-        self._entries = []
+        self.seen: set[tuple[str, str]] = set()
+        self.cache: dict[tuple[str, str], dict[str, str]] = {}
         self._integration_matchers = integration_matchers
-        self._description_cache = {}
+        self._cancel_scan: Callable[[], None] | None = None
+        self._ssdp_listeners: list[SSDPListener] = []
+        self._callbacks: list[tuple[Callable[[dict], None], dict[str, str]]] = []
+        self.flow_dispatcher: FlowDispatcher | None = None
+        self.description_manager: DescriptionManager | None = None
 
-    async def _on_ssdp_response(self, data: Mapping[str, Any]) -> None:
+    async def _async_on_ssdp_response(self, data: Mapping[str, Any]) -> None:
         """Process an ssdp response."""
-        self.async_store_entry(
+        await self._async_process_entry(
             ssdp.UPNPEntry({key.lower(): item for key, item in data.items()})
```

Comment:
If the only reason that we still rely on `netdisco` is `UPNPEntry`, we should consider copying that into this integration and _finally_ dropping netdisco.

Reply:
We still need `netdisco` for `discovery`, but at least we could drop it here.

Reply:
Dropped in the last two commits